From 9e4a222527b25210136c4b5ab48de9b70c187d87 Mon Sep 17 00:00:00 2001
From: Hannes Frederic Sowa <hannes@redhat.com>
Date: Tue, 7 Oct 2014 20:58:08 -0400
Subject: [net] ipv6: bump genid when delete/add address

Message-id: <791c48fe77eeacd96e42b12550c0dccdfeab5031.1412712186.git.hannes@redhat.com>
Patchwork-id: 97144
O-Subject: [RHEL7.1 net PATCH v2 4/6] ipv6: bump genid when delete/add address
Bugzilla: 1150017
RH-Acked-by: Jiri Benc <jbenc@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>
RH-Acked-by: Daniel Borkmann <dborkman@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1150017
Upstream Status: 439677d766ba9095e5afc4a30147f65bc363b6e7
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=8075311

Upstream Commit:
commit 439677d766ba9095e5afc4a30147f65bc363b6e7
Author: fan.du <fan.du@windriver.com>
Date:   Thu Aug 1 17:44:44 2013 +0800

    ipv6: bump genid when delete/add address

    Server           Client
    2001:1::803/64  <-> 2001:1::805/64
    2001:2::804/64  <-> 2001:2::806/64

    Server side fib binary tree looks like this:

                                       (2001:/64)
                                       /
                                      /
                       ffff88002103c380
                     /                 \
         (2)        /                   \
     (2001::803/128)                     ffff880037ac07c0
                                        /               \
                                       /                 \  (3)
                          ffff880037ac0640               (2001::806/128)
                           /             \
                 (1)      /               \
            (2001::804/128)               (2001::805/128)

    Delete 2001::804/64 won't cause prefix route deleted as well as rt in (3)
    destinate to 2001::806 with source address as 2001::804/64. That's because
    2001::803/64 is still alive, which make onlink=1 in ipv6_del_addr, this is
    where the substantial difference between same prefix configuration and
    different prefix configuration :) So packet are still transmitted out to
    2001::806 with source address as 2001::804/64.

    So bump genid will clear rt in (3), and up layer protocol will eventually
    find the right one for themselves.

    This problem arised from the discussion in here:
    http://marc.info/?l=linux-netdev&m=137404469219410&w=4

    Signed-off-by: Fan Du <fan.du@windriver.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Hannes Frederic Sowa <hannes@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv6/addrconf.c b/net/ipv6/addrconf.c
index 12f48eb..81dfed4 100644
--- a/net/ipv6/addrconf.c
+++ b/net/ipv6/addrconf.c
@@ -4708,6 +4708,7 @@ static void __ipv6_ifa_notify(int event, struct inet6_ifaddr *ifp)
   break;
  }
  atomic_inc(&net->ipv6.dev_addr_genid);
+ rt_genid_bump_ipv6(net);
 }
 
 static void ipv6_ifa_notify(int event, struct inet6_ifaddr *ifp)
-- 
1.7.1